core/src/provider.rs

```diff
@@ -54,46 +93,218 @@ impl JsonRpcCachedProvider {
         Ok(None)
     }
 
-    pub async fn get_block_number(&self) -> Result<U64, ProviderError> {
+    async fn get_chain_id(&self) -> Result<U256, ProviderError> {
+        self.provider.get_chainid().await
+    }
+
+    async fn get_block_number(&self) -> Result<U64, ProviderError> {
         self.provider.get_block_number().await
     }
 
-    pub async fn get_logs(&self, filter: &RindexerEventFilter) -> Result<Vec<Log>, ProviderError> {
+    async fn get_logs(&self, filter: &RindexerEventFilter) -> Result<Vec<Log>, ProviderError> {
         self.provider.get_logs(filter.raw_filter()).await
     }
 
-    pub async fn get_chain_id(&self) -> Result<U256, ProviderError> {
-        self.provider.get_chainid().await
+    fn max_block_range(&self) -> Option<U64> {
+        self.max_block_range
     }
 
-    pub fn get_inner_provider(&self) -> Arc<Provider<RetryClient<Http>>> {
-        Arc::clone(&self.provider)
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
+#[derive(Debug, Clone)]
+pub struct HyperSyncProvider {
+    provider: Arc<Client>,
+    pub max_block_range: Option<U64>,
+}
+
+impl HyperSyncProvider {
+    pub fn new(provider: Client, max_block_range: Option<U64>) -> Self {
+        HyperSyncProvider { provider: Arc::new(provider), max_block_range }
+    }
+}
+
+#[async_trait]
+impl ProviderInterface for HyperSyncProvider {
+    async fn get_latest_block(&self) -> Result<Option<Arc<Block<H256>>>, ProviderError> {
+        let latest_block = self
+            .provider
+            .get_height()
+            .await
+            .map_err(|err| ProviderError::CustomError(err.to_string()))?;
+
+        let query = blocks_and_transactions(latest_block, Some(latest_block + 1));
+
+        let data = self
+            .provider
+            .clone()
+            .collect(query, Default::default())
+            .await
+            .map_err(|err| ProviderError::CustomError(err.to_string()))?;
+
+        let block = data.data.blocks.iter().flatten().last().cloned();
+        let txs = data.data.transactions.into_iter().flatten().collect::<Vec<_>>();
+
+        block
+            .map(|b| {
+                b.try_into_ethers_hash(txs.into_iter().filter_map(|tx| tx.hash).collect())
+                    .map(Arc::new)
+            })
+            .transpose()
+            .map_err(|err| ProviderError::CustomError(err.to_string()))
+    }
+
+    async fn get_chain_id(&self) -> Result<U256, ProviderError> {
+        self.provider
+            .get_chain_id()
+            .await
+            .map(U256::from)
+            .map_err(|err| ProviderError::CustomError(err.to_string()))
+    }
+
+    async fn get_block_number(&self) -> Result<U64, ProviderError> {
+        self.provider
+            .get_height()
+            .await
+            .map(U64::from)
+            .map_err(|err| ProviderError::CustomError(err.to_string()))
+    }
+
+    async fn get_logs(&self, filter: &RindexerEventFilter) -> Result<Vec<Log>, ProviderError> {
```

Comment:
i did comment this but worth looking at this https://github.com/joshstevens19/rindexer/commit/d1683fbeeaabf4313424f81708193f81908c5f1f as we exposed a new log type to expose the block timestamp